{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d014bfb3-225e-4eac-82e7-f9cae7492fa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Old packages: 'AnnotationDbi', 'AnnotationHub', 'BiocParallel', 'Biostrings',\n",
      "  'caret', 'data.table', 'desc', 'fansi', 'GenomeInfoDb', 'googlesheets4',\n",
      "  'knitr', 'later', 'MatrixGenerics', 'mime', 'rcmdcheck', 'RcppArmadillo',\n",
      "  'readr', 'recipes', 'remotes', 'Rhdf5lib', 'RSQLite', 'S4Vectors', 'shiny',\n",
      "  'tibble', 'tidymodels', 'tidyr', 'tinytex', 'xfun'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'TxDb.Hsapiens.UCSC.hg19.knownGene'”\n",
      "Old packages: 'AnnotationDbi', 'AnnotationHub', 'BiocParallel', 'Biostrings',\n",
      "  'caret', 'data.table', 'desc', 'fansi', 'GenomeInfoDb', 'googlesheets4',\n",
      "  'knitr', 'later', 'MatrixGenerics', 'mime', 'rcmdcheck', 'RcppArmadillo',\n",
      "  'readr', 'recipes', 'remotes', 'Rhdf5lib', 'RSQLite', 'S4Vectors', 'shiny',\n",
      "  'tibble', 'tidymodels', 'tidyr', 'tinytex', 'xfun'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'rtracklayer'”\n",
      "Old packages: 'AnnotationDbi', 'AnnotationHub', 'BiocParallel', 'Biostrings',\n",
      "  'caret', 'data.table', 'desc', 'fansi', 'GenomeInfoDb', 'googlesheets4',\n",
      "  'knitr', 'later', 'MatrixGenerics', 'mime', 'rcmdcheck', 'RcppArmadillo',\n",
      "  'readr', 'recipes', 'remotes', 'Rhdf5lib', 'RSQLite', 'S4Vectors', 'shiny',\n",
      "  'tibble', 'tidymodels', 'tidyr', 'tinytex', 'xfun'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'GenomicRanges'”\n",
      "Old packages: 'AnnotationDbi', 'AnnotationHub', 'BiocParallel', 'Biostrings',\n",
      "  'caret', 'data.table', 'desc', 'fansi', 'GenomeInfoDb', 'googlesheets4',\n",
      "  'knitr', 'later', 'MatrixGenerics', 'mime', 'rcmdcheck', 'RcppArmadillo',\n",
      "  'readr', 'recipes', 'remotes', 'Rhdf5lib', 'RSQLite', 'S4Vectors', 'shiny',\n",
      "  'tibble', 'tidymodels', 'tidyr', 'tinytex', 'xfun'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 0 ranges and 0 metadata columns:\n",
       "   seqnames    ranges strand\n",
       "      <Rle> <IRanges>  <Rle>\n",
       "  -------\n",
       "  seqinfo: no sequences"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 0 ranges and 0 metadata columns:\n",
       "   seqnames    ranges strand\n",
       "      <Rle> <IRanges>  <Rle>\n",
       "  -------\n",
       "  seqinfo: no sequences"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "ERROR",
     "evalue": "Error in `[[<-`(`*tmp*`, name, value = \"-\"): 1 elements in value to replace 0 elements\n",
     "output_type": "error",
     "traceback": [
      "Error in `[[<-`(`*tmp*`, name, value = \"-\"): 1 elements in value to replace 0 elements\nTraceback:\n",
      "1. `$<-`(`*tmp*`, name, value = \"-\")",
      "2. `$<-`(`*tmp*`, name, value = \"-\")",
      "3. `[[<-`(`*tmp*`, name, value = \"-\")",
      "4. `[[<-`(`*tmp*`, name, value = \"-\")",
      "5. stop(paste(lv, \"elements in value to replace\", nrx, \"elements\"))"
     ]
    }
   ],
   "source": [
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "    BiocManager::install(version = \"3.13\")\n",
    "\n",
    "BiocManager::install(\"TxDb.Hsapiens.UCSC.hg19.knownGene\")\n",
    "BiocManager::install(\"rtracklayer\")\n",
    "BiocManager::install(\"GenomicRanges\")\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "test_path <- system.file(\"home/jovyan/work/imr_grohmmfinal.transcripts.bed\", package = \"rtracklayer\")\n",
    "test_bed <- file.path(test_path, \"home/jovyan/work/imr_grohmmfinal.transcripts.bed\")\n",
    "test <- import(test_bed)\n",
    "test\n",
    "test_bed_file <- BEDFile(test_bed)\n",
    "import(test_bed_file)\n",
    "test_bed_con <- file(test_bed)\n",
    "\n",
    "#groHMM data (starting with bed file of txs)\n",
    "\n",
    "#create GRanges of annotated regions in hg19\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "library(TxDb.Hsapiens.UCSC.hg19.knownGene)\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene\n",
    "tx <- transcripts(txdb)  \n",
    "\n",
    "#read in final tx bed files from groHMM\n",
    "txIMR <- rtracklayer::import(\"imr_grohmmfinal.transcripts.bed\")\n",
    "txIMR$name <- paste0(seqnames(txIMR),\"-\" , start(txIMR), strand(txIMR))\n",
    "txGM <- rtracklayer::import(\"gm_grohmmfinal.transcripts.bed\")\n",
    "txGM$name <- paste0(seqnames(txGM),\"-\" , start(txGM), strand(txGM))\n",
    "\n",
    "#subset TX list by intergenic regions\n",
    "txIMR_int <- subsetByOverlaps(txIMR, tx, type = \"any\", invert = T)\n",
    "txGM_int <- subsetByOverlaps(txGM, tx, type = \"any\", invert = T)\n",
    "\n",
    "#further subset by overlap with histone marks\n",
    "#create GRanges for histone bed files\n",
    "#to make import work, I had to delete the \"score\" column (5) and chrM entries in the IMR90 bed files\n",
    "bed <- read.table(\"GSM469967_UCSD.IMR90.H3K27ac.LL235.bed.gz\")\n",
    "bed$V5 <- NULL\n",
    "bed <- filter(bed, V1 != \"chrM\")\n",
    "write.table(bed, \"GSM469967_UCSD.IMR90.H3K27ac.LL235.bed\", quote = F, sep=\"\\t\", row.names=F, col.names=F)\n",
    "\n",
    "bed <- read.table(\"GSM521898_UCSD.IMR90.H3K4me1.SYL133.bed.gz\")\n",
    "bed$V5 <- NULL\n",
    "bed <- filter(bed, V1 != \"chrM\")\n",
    "write.table(bed, \"GSM521898_UCSD.IMR90.H3K4me1.SYL133.bed\", quote = F, sep=\"\\t\", row.names=F, col.names=F)\n",
    "rm(bed)\n",
    "\n",
    "IMR_H3K27ac <- rtracklayer::import(\"GSM469967_UCSD.IMR90.H3K27ac.LL235.bed\",format = \"bed\")\n",
    "IMR_H3K4me1 <- rtracklayer::import(\"GSM521898_UCSD.IMR90.H3K4me1.SYL133.bed\",format = \"bed\")\n",
    "\n",
    "extraCols <- c(signalValue = \"numeric\", pValue = \"numeric\", qValue = \"numeric\")\n",
    "GM_H3K4me1 <- rtracklayer::import(\"GSM733772_GM_H3K4me1.broadPeak\",format = \"bed\", extraCols = extraCols)\n",
    "GM_H3K27ac <- rtracklayer::import(\"GSM733771_GM_H3K27ac.broadPeak\",format = \"bed\", extraCols = extraCols)\n",
    "\n",
    "txIMR_eRNA <- subsetByOverlaps(txIMR_int, IMR_H3K27ac)\n",
    "txIMR_eRNA <- subsetByOverlaps(txIMR_eRNA, IMR_H3K4me1)\n",
    "export(txIMR_eRNA, \"IMR_eRNA.gtf\")\n",
    "\n",
    "txGM_eRNA <- subsetByOverlaps(txGM_int, GM_H3K27ac)\n",
    "txGM_eRNA <- subsetByOverlaps(txGM_eRNA, GM_H3K4me1)\n",
    "export(txGM_eRNA, \"GM_eRNA.gtf\")\n",
    "\n",
    "#feature counts from groHMM results\n",
    "library(Rsubread)\n",
    "bamsIMR <- list.files( \"./bam_files\", pattern=glob2rx(\"*IMR*bam$\"), full=TRUE )\n",
    "bamsGM <- list.files( \"./bam_files\", pattern=glob2rx(\"*GM*bam$\"), full=TRUE )\n",
    "geneCtsIMR <- featureCounts(files = bamsIMR, annot.inbuilt = \"hg19\", useMetaFeatures = T, strandSpecific = 1, nthreads = 8)\n",
    "geneCtsGM <- featureCounts(files = bamsGM, annot.inbuilt = \"hg19\", useMetaFeatures = T, strandSpecific = 1, nthreads = 8)\n",
    "eRNACtsGM <- featureCounts(files = bamsGM, annot.ext = \"GM_eRNA.gtf\", isGTFAnnotationFile = T, GTF.featureType = \"sequence_feature\", GTF.attrType = \"name\", strandSpecific = 1, nthreads = 8)\n",
    "eRNACtsIMR <- featureCounts(files = bamsIMR, annot.ext = \"IMR_eRNA.gtf\", isGTFAnnotationFile = T, GTF.featureType = \"sequence_feature\", GTF.attrType = \"name\", strandSpecific = 1, nthreads = 8)\n",
    "\n",
    "#create count matrix\n",
    "#geneCts[4] provides quality/mapping info\n",
    "IMRc1 <- as.data.frame(geneCtsIMR[1])\n",
    "IMRc2 <- as.data.frame(geneCtsIMR[2])\n",
    "mRNAimr <- cbind(IMRc1, IMRc2)\n",
    "\n",
    "IMRc1 <- as.data.frame(eRNACtsIMR[1])\n",
    "IMRc2 <- as.data.frame(eRNACtsIMR[2])\n",
    "eRNAimr <- cbind(IMRc1, IMRc2)\n",
    "rm(IMRc1, IMRc2)\n",
    "\n",
    "GMc1 <- as.data.frame(geneCtsGM[1])\n",
    "GMc2 <- as.data.frame(geneCtsGM[2])\n",
    "mRNAgm <- cbind(GMc1, GMc2)\n",
    "\n",
    "GMc1 <- as.data.frame(eRNACtsGM[1])\n",
    "GMc2 <- as.data.frame(eRNACtsGM[2])\n",
    "eRNAgm <- cbind(GMc1, GMc2)\n",
    "rm(GMc1, GMc2)\n",
    "\n",
    "#clean up count files\n",
    "library(stringr)\n",
    "colnames(mRNAimr)[13:18] <- sub(\"annotation.\", \"\", colnames(mRNAimr)[13:18])\n",
    "colnames(mRNAimr)[1:12] <- sub(c(\"counts.\"), \"\", colnames(mRNAimr)[1:12])\n",
    "colnames(mRNAimr)[1:12] <- sub(c(\".bam\"), \"\", colnames(mRNAimr)[1:12])\n",
    "\n",
    "colnames(mRNAgm)[13:18] <- sub(\"annotation.\", \"\", colnames(mRNAgm)[13:18])\n",
    "colnames(mRNAgm)[1:12] <- sub(c(\"counts.\"), \"\", colnames(mRNAgm)[1:12])\n",
    "colnames(mRNAgm)[1:12] <- sub(c(\".bam\"), \"\", colnames(mRNAgm)[1:12])\n",
    "\n",
    "colnames(eRNAimr)[13:18] <- sub(\"annotation.\", \"\", colnames(eRNAimr)[13:18])\n",
    "colnames(eRNAimr)[1:12] <- sub(c(\"counts.\"), \"\", colnames(eRNAimr)[1:12])\n",
    "colnames(eRNAimr)[1:12] <- sub(c(\".bam\"), \"\", colnames(eRNAimr)[1:12])\n",
    "\n",
    "colnames(eRNAgm)[13:18] <- sub(\"annotation.\", \"\", colnames(eRNAgm)[13:18])\n",
    "colnames(eRNAgm)[1:12] <- sub(c(\"counts.\"), \"\", colnames(eRNAgm)[1:12])\n",
    "colnames(eRNAgm)[1:12] <- sub(c(\".bam\"), \"\", colnames(eRNAgm)[1:12])\n",
    "\n",
    "mRNAimr$Chr <- word(mRNAimr$Chr, 1, sep = \"\\\\;\")\n",
    "mRNAimr$Start <- word(mRNAimr$Start, 1, sep = \"\\\\;\")\n",
    "mRNAimr$End <- word(mRNAimr$End, 1, sep = \"\\\\;\")\n",
    "mRNAimr$Strand <- word(mRNAimr$Strand, 1, sep = \"\\\\;\")\n",
    "mRNAimr$Start <- as.numeric(mRNAimr$Start)\n",
    "mRNAimr$End <- as.numeric(mRNAimr$End)\n",
    "\n",
    "mRNAgm$Chr <- word(mRNAgm$Chr, 1, sep = \"\\\\;\")\n",
    "mRNAgm$Start <- word(mRNAgm$Start, 1, sep = \"\\\\;\")\n",
    "mRNAgm$End <- word(mRNAgm$End, 1, sep = \"\\\\;\")\n",
    "mRNAgm$Strand <- word(mRNAgm$Strand, 1, sep = \"\\\\;\")\n",
    "mRNAgm$Start <- as.numeric(mRNAgm$Start)\n",
    "mRNAgm$End <- as.numeric(mRNAgm$End)\n",
    "\n",
    "eRNAimr$Chr <- word(eRNAimr$Chr, 1, sep = \"\\\\;\")\n",
    "eRNAimr$Start <- word(eRNAimr$Start, 1, sep = \"\\\\;\")\n",
    "eRNAimr$End <- word(eRNAimr$End, 1, sep = \"\\\\;\")\n",
    "eRNAimr$Strand <- word(eRNAimr$Strand, 1, sep = \"\\\\;\")\n",
    "eRNAimr$Start <- as.numeric(eRNAimr$Start)\n",
    "eRNAimr$End <- as.numeric(eRNAimr$End)\n",
    "\n",
    "eRNAgm$Chr <- word(eRNAgm$Chr, 1, sep = \"\\\\;\")\n",
    "eRNAgm$Start <- word(eRNAgm$Start, 1, sep = \"\\\\;\")\n",
    "eRNAgm$End <- word(eRNAgm$End, 1, sep = \"\\\\;\")\n",
    "eRNAgm$Strand <- word(eRNAgm$Strand, 1, sep = \"\\\\;\")\n",
    "eRNAgm$Start <- as.numeric(eRNAgm$Start)\n",
    "eRNAgm$End <- as.numeric(eRNAgm$End)\n",
    "\n",
    "#filter by protein coding genes (for mRNA counts)\n",
    "library(biomaRt)\n",
    "library(dplyr)\n",
    "ensembl <- useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "gb <- getBM(attributes = c(\"hgnc_symbol\", \"gene_biotype\", \"entrezgene_id\"), mart = ensembl)\n",
    "colnames(gb)[3] <- \"GeneID\"\n",
    "mRNAimr <- left_join(mRNAimr, gb, by = \"GeneID\")\n",
    "mRNAimr <- subset(mRNAimr, gene_biotype == \"protein_coding\")\n",
    "mRNAimr$gene_biotype <- NULL\n",
    "mRNAimr <- distinct(mRNAimr, GeneID, .keep_all = T)\n",
    "\n",
    "mRNAgm <- left_join(mRNAgm, gb, by = \"GeneID\")\n",
    "mRNAgm <- subset(mRNAgm, gene_biotype == \"protein_coding\")\n",
    "mRNAgm$gene_biotype <- NULL\n",
    "mRNAgm <- distinct(mRNAgm, GeneID, .keep_all = T)\n",
    "\n",
    "#combine regions of overlapping fragments or fragments within 500 bp of each other\n",
    "library(GenomicRanges)\n",
    "e_regionsIMR <- eRNAimr[, 13:17]\n",
    "e_regionsIMR$Start <- e_regionsIMR$Start - 250\n",
    "e_regionsIMR$End <- e_regionsIMR$End + 250\n",
    "GR <- makeGRangesFromDataFrame(e_regionsIMR, keep.extra.columns = T)\n",
    "e_regionsIMR <- reduce(GR, ignore.strand = F)\n",
    "e_regionsIMR <- as.data.frame(e_regionsIMR)\n",
    "e_regionsIMR$start <- e_regionsIMR$start + 250\n",
    "e_regionsIMR$end <- e_regionsIMR$end - 250\n",
    "\n",
    "e_regionsGM <- eRNAgm[, 13:17]\n",
    "e_regionsGM$Start <- e_regionsGM$Start - 250\n",
    "e_regionsGM$End <- e_regionsGM$End + 250\n",
    "GR <- makeGRangesFromDataFrame(e_regionsGM, keep.extra.columns = T)\n",
    "e_regionsGM <- reduce(GR, ignore.strand = F)\n",
    "e_regionsGM <- as.data.frame(e_regionsGM)\n",
    "e_regionsGM$start <- e_regionsGM$start + 250\n",
    "e_regionsGM$end <- e_regionsGM$end - 250\n",
    "\n",
    "#combining and renaming all eRNAs between GM and IMR\n",
    "library(GenomicRanges)\n",
    "all_eRNA <- rbind(e_regionsIMR, e_regionsGM)\n",
    "all_eRNA$Geneid <- NULL\n",
    "GR <- makeGRangesFromDataFrame(all_eRNA, keep.extra.columns = T)\n",
    "all_eRNA <- reduce(GR, ignore.strand = F)\n",
    "all_eRNA <- as.data.frame(all_eRNA)\n",
    "all_eRNA$width <- NULL\n",
    "all_eRNA$seq <- ave(all_eRNA$start, all_eRNA$seqnames, FUN = seq_along)\n",
    "all_eRNA$Geneid <- paste0(all_eRNA$seqnames, \"-\", all_eRNA$seq, all_eRNA$strand)\n",
    "\n",
    "#map combined regions back to original count file\n",
    "GR <- makeGRangesFromDataFrame(eRNAimr, keep.extra.columns = T)\n",
    "GR1 <- makeGRangesFromDataFrame(all_eRNA, keep.extra.columns = T)\n",
    "pairs <- mergeByOverlaps(GR, GR1, type = \"any\", select = \"all\", ignore.strand = F)\n",
    "pairs1 <- data.frame(matrix(NA, nrow = nrow(pairs[1]), ncol = 2))\n",
    "pairs1[1] <- pairs[[14]]\n",
    "pairs1[2] <- pairs[[18]]\n",
    "colnames(pairs1) <- c(\"GeneID\", \"region\")\n",
    "eRNAimr <- left_join(eRNAimr, pairs1, by = \"GeneID\")\n",
    "\n",
    "GR <- makeGRangesFromDataFrame(eRNAgm, keep.extra.columns = T)\n",
    "pairs <- mergeByOverlaps(GR, GR1, type = \"any\", select = \"all\", ignore.strand = F)\n",
    "pairs1 <- data.frame(matrix(NA, nrow = nrow(pairs[1]), ncol = 2))\n",
    "pairs1[1] <- pairs[[14]]\n",
    "pairs1[2] <- pairs[[18]]\n",
    "colnames(pairs1) <- c(\"GeneID\", \"region\")\n",
    "eRNAgm <- left_join(eRNAgm, pairs1, by = \"GeneID\")\n",
    "\n",
    "#provide GeneHancer ID for each annotated eRNA by overlap\n",
    "library(GenomicRanges)\n",
    "GH_HG19 <- readRDS(\"GH_HG19.rds\")\n",
    "GR_GH <- makeGRangesFromDataFrame(GH_HG19, keep.extra.columns = T)\n",
    "GR <- makeGRangesFromDataFrame(eRNAimr, keep.extra.columns = T)\n",
    "GHpairs <- mergeByOverlaps(GR, GR_GH, type = \"any\", select = \"all\", ignore.strand = T)\n",
    "GH_pairsIMR <- data.frame(matrix(NA, nrow = nrow(GHpairs[1]), ncol = 2))\n",
    "colnames(GH_pairsIMR) <- c(\"GeneID\", \"GeneHancerID\")\n",
    "GH_pairsIMR$GeneID <- GHpairs[[14]]\n",
    "GH_pairsIMR$GeneHancerID <- GHpairs[[22]]\n",
    "\n",
    "GR <- makeGRangesFromDataFrame(eRNAgm, keep.extra.columns = T)\n",
    "GHpairs <- mergeByOverlaps(GR, GR_GH, type = \"any\", select = \"all\", ignore.strand = T)\n",
    "GH_pairsGM <- data.frame(matrix(NA, nrow = nrow(GHpairs[1]), ncol = 2))\n",
    "colnames(GH_pairsGM) <- c(\"GeneID\", \"GeneHancerID\")\n",
    "GH_pairsGM$GeneID <- GHpairs[[14]]\n",
    "GH_pairsGM$GeneHancerID <- GHpairs[[22]]\n",
    "\n",
    "#add GH ID to count file\n",
    "eRNAimr <- left_join(eRNAimr, GH_pairsIMR, by = \"GeneID\")\n",
    "eRNAimr$GeneHancerID <- ifelse(is.na(eRNAimr$GeneHancerID), NA , paste0(eRNAimr$GeneHancerID, eRNAimr$Strand))\n",
    "\n",
    "eRNAgm <- left_join(eRNAgm, GH_pairsGM, by = \"GeneID\")\n",
    "eRNAgm$GeneHancerID <- ifelse(is.na(eRNAgm$GeneHancerID), NA , paste0(eRNAgm$GeneHancerID, eRNAgm$Strand))\n",
    "\n",
    "#rename groHMM tx to new GH ID or region ID\n",
    "eRNAimr$GeneID <- ifelse(is.na(eRNAimr$GeneHancerID), eRNAimr$region, eRNAimr$GeneHancerID)\n",
    "eRNAimr[, 19:21] <- NULL\n",
    "\n",
    "eRNAgm$GeneID <- ifelse(is.na(eRNAgm$GeneHancerID), eRNAgm$region, eRNAgm$GeneHancerID)\n",
    "eRNAgm[, 19:21] <- NULL\n",
    "\n",
    "#collapse and sum counts by region\n",
    "df1 <- eRNAimr[, 13:17]\n",
    "df1 <- df1 %>% group_by(GeneID, Chr, Strand) %>% summarize(Start = min(Start), End = max(End))\n",
    "df1$Length <- df1$End - df1$Start\n",
    "df2 <- eRNAimr[, 1:13]\n",
    "df2 <- df2 %>% group_by(GeneID) %>% summarize_each(function(x)sum(x))\n",
    "eRNAimr <- inner_join(df1, df2, by = \"GeneID\")\n",
    "eIMR_coord <- eRNAimr[, 1:6]\n",
    "rm(df1, df2)\n",
    "\n",
    "df1 <- eRNAgm[, 13:17]\n",
    "df1 <- df1 %>% group_by(GeneID, Chr, Strand) %>% summarize(Start = min(Start), End = max(End))\n",
    "df1$Length <- df1$End - df1$Start\n",
    "df2 <- eRNAgm[, 1:13]\n",
    "df2 <- df2 %>% group_by(GeneID) %>% summarize_each(function(x)sum(x))\n",
    "eRNAgm <- inner_join(df1, df2, by = \"GeneID\")\n",
    "eGM_coord <- eRNAgm[, 1:6]\n",
    "rm(df1, df2)\n",
    "\n",
    "#TPM calculation\n",
    "eIMRtpm <- eRNAimr\n",
    "eIMRtpm[,7:18] <- lapply(eIMRtpm[,7:18], function(x) ((x/(eIMRtpm$Length/1000))))\n",
    "eIMRtpm[,7:18] <- lapply(eIMRtpm[,7:18], function(x) ((x/(sum(x)/1000000))))\n",
    "#colSums(eIMRtpm[7:18])\n",
    "\n",
    "mIMRtpm <- mRNAimr\n",
    "mIMRtpm[,1:12] <- lapply(mIMRtpm[,1:12], function(x) ((x/(mIMRtpm$Length/1000))))\n",
    "mIMRtpm[,1:12] <- lapply(mIMRtpm[,1:12], function(x) ((x/(sum(x)/1000000))))\n",
    "#colSums(mIMRtpm[1:12])\n",
    "\n",
    "eGMtpm <- eRNAgm\n",
    "eGMtpm[,7:18] <- lapply(eGMtpm[,7:18], function(x) ((x/(eGMtpm$Length/1000))))\n",
    "eGMtpm[,7:18] <- lapply(eGMtpm[,7:18], function(x) ((x/(sum(x)/1000000))))\n",
    "#colSums(eGMtpm[7:18])\n",
    "\n",
    "mGMtpm <- mRNAgm\n",
    "mGMtpm[,1:12] <- lapply(mGMtpm[,1:12], function(x) ((x/(mGMtpm$Length/1000))))\n",
    "mGMtpm[,1:12] <- lapply(mGMtpm[,1:12], function(x) ((x/(sum(x)/1000000))))\n",
    "#colSums(mGMtpm[1:12])\n",
    "\n",
    "#reorder counts columns to be in time order\n",
    "eRNAimr <- eRNAimr[, c(1:7, 13, 10, 12, 15, 16, 18, 8, 9, 11, 14, 17)]\n",
    "mRNAimr <- mRNAimr[, c(13:19, 1,7,4,6,9,10,12,2,3,5,8,11)]\n",
    "eRNAgm <- eRNAgm[, c(1:7, 13, 10, 12, 15, 16, 18, 8, 9, 11, 14, 17)]\n",
    "mRNAgm <- mRNAgm[, c(13:19, 1,7,4,6,9,10,12,2,3,5,8,11)]\n",
    "\n",
    "\n",
    "#DESEQ2 NORMALIZATION OF COUNTS FOR DE ANALYSIS\n",
    "#count data less timepoints > 24H\n",
    "eIMRct <- eRNAimr[, c(1, 7:16)]\n",
    "eIMRct <- data.frame(eIMRct, row.names = 1)\n",
    "mIMRct <- mRNAimr[, c(1, 8:17)]\n",
    "mIMRct <- data.frame(mIMRct, row.names = 1)\n",
    "\n",
    "eGMct <- eRNAgm[, c(1, 7:16)]\n",
    "eGMct <- data.frame(eGMct, row.names = 1)\n",
    "mGMct <- mRNAgm[, c(1, 8:17)]\n",
    "mGMct <- data.frame(mGMct, row.names = 1)\n",
    "\n",
    "#sample files (required for DESeq object)\n",
    "colsIMR <- colnames(eIMRct)\n",
    "samps <- c(\"control\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\")\n",
    "coldataIMR <- data.frame(colsIMR, \"condition\" = samps)\n",
    "coldataIMR <- data.frame(coldataIMR, row.names = 1)\n",
    "\n",
    "colsGM <- colnames(eGMct)\n",
    "#samps <- c(\"control\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\", \"treated\")\n",
    "coldataGM <- data.frame(colsGM, \"condition\" = samps)\n",
    "coldataGM <- data.frame(coldataGM, row.names = 1)\n",
    "\n",
    "#convert to DESeq object and apply normalization - FOR DE ANALYSIS ONLY!\n",
    "library(DESeq2)\n",
    "dds_eIMR <- DESeqDataSetFromMatrix(countData = eIMRct, colData = coldataIMR, design = ~ condition)\n",
    "dds_eIMR <- estimateSizeFactors(dds_eIMR)\n",
    "eIMRnorm <- as.data.frame(counts(dds_eIMR, normalized = T))\n",
    "dds_mIMR <- DESeqDataSetFromMatrix(countData = mIMRct, colData = coldataIMR, design = ~ condition)\n",
    "dds_mIMR <- estimateSizeFactors(dds_mIMR)\n",
    "mIMRnorm <- as.data.frame(counts(dds_mIMR, normalized = T))\n",
    "\n",
    "dds_eGM <- DESeqDataSetFromMatrix(countData = eGMct, colData = coldataGM, design = ~ condition)\n",
    "dds_eGM <- estimateSizeFactors(dds_eGM)\n",
    "eGMnorm <- as.data.frame(counts(dds_eGM, normalized = T))\n",
    "dds_mGM <- DESeqDataSetFromMatrix(countData = mGMct, colData = coldataGM, design = ~ condition)\n",
    "dds_mGM <- estimateSizeFactors(dds_mGM)\n",
    "mGMnorm <- as.data.frame(counts(dds_mGM, normalized = T))\n",
    "\n",
    "#remove zeros and calculate |FC| based on min/max for each gene. \n",
    "#also filter for FC > 1\n",
    "#DOUBLE CHECK SUBSETS WITH GROHMM DATA\n",
    "eIMRnorm[eIMRnorm == 0] <- NA\n",
    "eIMRnorm$L2FC <- log2((do.call(pmax, c(eIMRnorm[, 1:10], na.rm = T)))/(do.call(pmin, c(eIMRnorm[, 1:10], na.rm = T))))\n",
    "eIMRnorm <- subset(eIMRnorm,L2FC > 1)\n",
    "mIMRnorm[mIMRnorm == 0] <- NA\n",
    "mIMRnorm$L2FC <- log2((do.call(pmax, c(mIMRnorm[, 1:10], na.rm = T)))/(do.call(pmin, c(mIMRnorm[, 1:10], na.rm = T))))\n",
    "mIMRnorm <- subset(mIMRnorm,L2FC > 1)\n",
    "\n",
    "eGMnorm[eGMnorm == 0] <- NA\n",
    "eGMnorm$L2FC <- log2((do.call(pmax, c(eGMnorm[, 1:10], na.rm = T)))/(do.call(pmin, c(eGMnorm[, 1:10], na.rm = T))))\n",
    "eGMnorm <- subset(eGMnorm,L2FC > 1)\n",
    "mGMnorm[mGMnorm == 0] <- NA\n",
    "mGMnorm$L2FC <- log2((do.call(pmax, c(mGMnorm[, 1:10], na.rm = T)))/(do.call(pmin, c(mGMnorm[, 1:10], na.rm = T))))\n",
    "mGMnorm <- subset(mGMnorm,L2FC > 1)\n",
    "\n",
    "#determine up or down-regulated based on slope of linear regression line\n",
    "slope <- function(x) {\n",
    "  tempDF <- data.frame(x, time = 7:18)\n",
    "  lm(x ~ time, data = tempDF, na.action = na.exclude)$coefficients[2]\n",
    "}\n",
    "\n",
    "#these all need to be double checked for the groHMM data\n",
    "tdata <- as.data.frame(t(eIMRnorm))\n",
    "eIMRnorm$slope <- sapply(tdata[7:18,], slope) \n",
    "eIMRnorm$DE <- ifelse(eIMRnorm$slope > 0, \"UP\", \"DOWN\")\n",
    "eIMR_DEG <- eIMRnorm[, c(1, 11, 13)]\n",
    "tdata <- as.data.frame(t(mIMRnorm))\n",
    "mIMRnorm$slope <- sapply(tdata[7:18,], slope)\n",
    "mIMRnorm$DE <- ifelse(mIMRnorm$slope > 0, \"UP\", \"DOWN\")\n",
    "\n",
    "tdata <- as.data.frame(t(eGMnorm))\n",
    "eGMnorm$slope <- sapply(tdata[7:18,], slope) \n",
    "eGMnorm$DE <- ifelse(eGMnorm$slope > 0, \"UP\", \"DOWN\")\n",
    "eGM_DEG <- eGMnorm[, c(1, 11, 13)]\n",
    "tdata <- as.data.frame(t(mGMnorm))\n",
    "mGMnorm$slope <- sapply(tdata[7:18,], slope)\n",
    "mGMnorm$DE <- ifelse(mGMnorm$slope > 0, \"UP\", \"DOWN\")\n",
    "rm(tdata)\n",
    "\n",
    "#continuity index\n",
    "#Peng's script for continuity index\n",
    "#this function works for smaller datasets, if you get a memory error, use function below\n",
    "CI <- function(x) {\n",
    "  return(diag(cor(t(x[, -ncol(x)]), t(x[, -1]), method = \"spearman\")))\n",
    "}\n",
    "\n",
    "#CI for very large datasets\n",
    "CI2 <- function(x) {\n",
    "  tx1 <- t(x[, -ncol(x)])\n",
    "  tx2 <- t(x[, -1])\n",
    "  cors <- c()\n",
    "  for(i in 1:ncol(tx1)){\n",
    "    c <- cor(tx1[, i], tx2[, i], method = \"spearman\")\n",
    "    cors <- c(cors, c)\n",
    "  }\n",
    "  return(cors)\n",
    "}\n",
    "\n",
    "#only test DEGs and DEEs\n",
    "library(tidyr)\n",
    "eIMRnorm$Geneid <- row.names(eIMRnorm)\n",
    "eIMRnorm_SE <- separate_rows(eIMRnorm, Geneid, sep = \",\")\n",
    "eIMRnorm_SE <- distinct(eIMRnorm_SE, Geneid, .keep_all = T)\n",
    "eIMRnorm_SE[is.na(eIMRnorm_SE)] <- 0\n",
    "eIMRnorm_SE$CI <- CI2(eIMRnorm_SE[, 1:10])\n",
    "ci_eIMR <- subset(eIMRnorm_SE, eIMRnorm_SE$CI > 0.2 | eIMRnorm_SE$CI < -0.2)\n",
    "static_eIMR <- anti_join(eIMRnorm_SE, ci_eIMR)\n",
    "\n",
    "mIMRnorm$Geneid <- row.names(mIMRnorm)\n",
    "mIMRnorm[is.na(mIMRnorm)] <- 0\n",
    "mIMRnorm$CI <- CI(mIMRnorm[, 1:10])\n",
    "ci_mIMR <- subset(mIMRnorm, mIMRnorm$CI > 0.2 | mIMRnorm$CI < -0.2)\n",
    "\n",
    "ieIMR <- subset(eIMR_coord, GeneID %in% ci_eIMR$Geneid) ##coordinates for all inducible eRNAs\n",
    "seIMR <- subset(eIMR_coord, GeneID %in% static_eIMR$Geneid) ##coordinates for all static eRNAs\n",
    "\n",
    "eGMnorm$Geneid <- row.names(eGMnorm)\n",
    "eGMnorm_SE <- separate_rows(eGMnorm, Geneid, sep = \",\")\n",
    "eGMnorm_SE <- distinct(eGMnorm_SE, Geneid, .keep_all = T)\n",
    "eGMnorm_SE[is.na(eGMnorm_SE)] <- 0\n",
    "eGMnorm_SE$CI <- CI2(eGMnorm_SE[, 1:10])\n",
    "ci_eGM <- subset(eGMnorm_SE, eGMnorm_SE$CI > 0.2 | eGMnorm_SE$CI < -0.2)\n",
    "static_eGM <- anti_join(eGMnorm_SE, ci_eGM)\n",
    "\n",
    "mGMnorm$Geneid <- row.names(mGMnorm)\n",
    "mGMnorm[is.na(mGMnorm)] <- 0\n",
    "mGMnorm$CI <- CI(mGMnorm[, 1:10])\n",
    "ci_mGM <- subset(mGMnorm, mGMnorm$CI > 0.2 | mGMnorm$CI < -0.2)\n",
    "\n",
    "ieGM <- subset(eGM_coord, GeneID %in% ci_eGM$Geneid) ##coordinates for all inducible eRNAs\n",
    "seGM <- subset(eGM_coord, GeneID %in% static_eGM$Geneid) ##coordinates for all static eRNAs\n",
    "\n",
    "#add gene symbol to mRNA lists\n",
    "gb$GeneID <- as.character(gb$GeneID)\n",
    "ci_mIMR <- left_join(ci_mIMR, gb[,c(1,3) ], by = c(\"Geneid\" = \"GeneID\"))\n",
    "ci_mGM <- left_join(ci_mGM, gb[,c(1,3) ], by = c(\"Geneid\" = \"GeneID\"))\n",
    "\n",
    "#find overlapping eRNA's for DE genes (EP PAIRS) - within 200kb\n",
    "#Downloaded GTF data from UCSC on Refseq genes for HG19\n",
    "library(rtracklayer)\n",
    "hg19gtf <- as.data.frame(rtracklayer::import(\"hg19.ncbiRefSeq.gtf\"))\n",
    "\n",
    "mIMR_DEG.gtf <- subset(hg19gtf, gene_id %in% ci_mIMR$hgnc_symbol)\n",
    "mIMR_DEG.gtf <- subset(mIMR_DEG.gtf, type == \"transcript\")\n",
    "mIMR_DEG.gtf <- mIMR_DEG.gtf[order(mIMR_DEG.gtf$gene_id, -mIMR_DEG.gtf$width),]\n",
    "mIMR_DEG.gtf <- distinct(mIMR_DEG.gtf, gene_id, .keep_all = T)\n",
    "\n",
    "exp_genesIMR <- mIMR_DEG.gtf\n",
    "exp_genesIMR$start <- exp_genesIMR$start - 200000\n",
    "exp_genesIMR$end <- exp_genesIMR$end + 200000\n",
    "GR1 <- makeGRangesFromDataFrame(exp_genesIMR, keep.extra.columns = T)\n",
    "GR2 <- makeGRangesFromDataFrame(ieIMR, keep.extra.columns = T)\n",
    "pairs <- mergeByOverlaps(GR1, GR2, type = \"any\", select = \"all\", ignore.strand = T)\n",
    "pairsIMR <- data.frame(matrix(NA, nrow = nrow(pairs[1]), ncol = 2))\n",
    "pairsIMR[1] <- pairs[[6]]\n",
    "pairsIMR[2] <- pairs[[12]]\n",
    "colnames(pairsIMR) <- c(\"gene\", \"eRNA\")\n",
    "pairsIMR_combo <- pairsIMR %>% group_by(gene) %>% mutate(rn = row_number()) %>% spread(rn, eRNA)\n",
    "pairsIMR_combo <- as.data.frame(pairsIMR_combo)\n",
    "\n",
    "mGM_DEG.gtf <- subset(hg19gtf, gene_id %in% ci_mGM$hgnc_symbol)\n",
    "mGM_DEG.gtf <- subset(mGM_DEG.gtf, type == \"transcript\")\n",
    "mGM_DEG.gtf <- mGM_DEG.gtf[order(mGM_DEG.gtf$gene_id, -mGM_DEG.gtf$width),]\n",
    "mGM_DEG.gtf <- distinct(mGM_DEG.gtf, gene_id, .keep_all = T)\n",
    "\n",
    "exp_genesGM <- mGM_DEG.gtf\n",
    "exp_genesGM$start <- exp_genesGM$start - 200000\n",
    "exp_genesGM$end <- exp_genesGM$end + 200000\n",
    "GR1 <- makeGRangesFromDataFrame(exp_genesGM, keep.extra.columns = T)\n",
    "GR2 <- makeGRangesFromDataFrame(ieGM, keep.extra.columns = T)\n",
    "pairs <- mergeByOverlaps(GR1, GR2, type = \"any\", select = \"all\", ignore.strand = T)\n",
    "pairsGM <- data.frame(matrix(NA, nrow = nrow(pairs[1]), ncol = 2))\n",
    "pairsGM[1] <- pairs[[6]]\n",
    "pairsGM[2] <- pairs[[12]]\n",
    "colnames(pairsGM) <- c(\"gene\", \"eRNA\")\n",
    "pairsGM_combo <- pairsGM %>% group_by(gene) %>% mutate(rn = row_number()) %>% spread(rn, eRNA)\n",
    "pairsGM_combo <- as.data.frame(pairsGM_combo)\n",
    "\n",
    "save.image(\"GM_IMR_gHMM.RData\")\n",
    "\n",
    "write.table(ci_mIMR, \"imIMR_hmm.txt\", sep = \"\\t\", col.names = T, row.names = F)\n",
    "write.table(ci_mGM, \"imGM_hmm.txt\", sep = \"\\t\", col.names = T, row.names = F)\n",
    "\n",
    "getwd()\n",
    "list.files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e99d649f-1f56-41b1-8bf8-c898a5e646cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No traceback available \n"
     ]
    }
   ],
   "source": [
    "traceback()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3555535e-2dae-4250-ad1d-d0f5545693f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/home/jovyan/work'"
      ],
      "text/latex": [
       "'/home/jovyan/work'"
      ],
      "text/markdown": [
       "'/home/jovyan/work'"
      ],
      "text/plain": [
       "[1] \"/home/jovyan/work\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09a9be02-dc31-4ef1-81ca-7c93634c16c1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
